lib/bb/runqueue.py

```diff
@@ -0,0 +1,3116 @@
+
+        max_process = int(self.cfgData.getVar("BB_NUMBER_PARSE_THREADS") or os.cpu_count() or 1)
+        # We cannot use the real multiprocessing.Pool easily due to some local data
+        # that can't be pickled. This is a cheap multi-process solution.
+        launched = []
+        while fns:
+            if len(launched) < max_process:
+                p = Process(target=self.rq_dump_sigfn, args=(fns.pop(), options))
+                p.start()
+                launched.append(p)
+            for q in launched:
+                # The finished processes are joined when calling is_alive()
+                if not q.is_alive():
+                    launched.remove(q)
+        for p in launched:
+                p.join()
+
+        bb.parse.siggen.dump_sigs(self.rqdata.dataCaches, options)
+
+        return
+
+    def print_diffscenetasks(self):
+
+        noexec = []
+        tocheck = set()
+
+        for tid in self.rqdata.runtaskentries:
+            (mc, fn, taskname, taskfn) = split_tid_mcfn(tid)
+            taskdep = self.rqdata.dataCaches[mc].task_deps[taskfn]
+
+            if 'noexec' in taskdep and taskname in taskdep['noexec']:
+                noexec.append(tid)
+                continue
+
+            tocheck.add(tid)
+
+        valid_new = self.validate_hashes(tocheck, self.cooker.data, 0, True, summary=False)
+
+        # Tasks which are both setscene and noexec never care about dependencies
+        # We therefore find tasks which are setscene and noexec and mark their
+        # unique dependencies as valid.
+        for tid in noexec:
+            if tid not in self.rqdata.runq_setscene_tids:
+                continue
+            for dep in self.rqdata.runtaskentries[tid].depends:
+                hasnoexecparents = True
+                for dep2 in self.rqdata.runtaskentries[dep].revdeps:
+                    if dep2 in self.rqdata.runq_setscene_tids and dep2 in noexec:
+                        continue
+                    hasnoexecparents = False
+                    break
+                if hasnoexecparents:
+                    valid_new.add(dep)
+
+        invalidtasks = set()
+        for tid in self.rqdata.runtaskentries:
+            if tid not in valid_new and tid not in noexec:
+                invalidtasks.add(tid)
+
+        found = set()
+        processed = set()
+        for tid in invalidtasks:
+            toprocess = set([tid])
+            while toprocess:
+                next = set()
+                for t in toprocess:
+                    for dep in self.rqdata.runtaskentries[t].depends:
+                        if dep in invalidtasks:
+                            found.add(tid)
+                        if dep not in processed:
+                            processed.add(dep)
+                            next.add(dep)
+                toprocess = next
+                if tid in found:
+                    toprocess = set()
+
+        tasklist = []
+        for tid in invalidtasks.difference(found):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To create a `list`, try to use `list` comprehension instead of a loop. List comprehension is the preferred way to make a list using Python, and it&#39;s simpler and easier to understand than using a loop.

[Learn more](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)